### ISSUE TITLE
Audio synchronisation questions.

### ISSUE THREAD

**Issue by audetto** [AppleWin/AppleWin#1220]
In the linux version, audio lag has always been an issue, but I have always struggled to follow the logic in Windows.

There are various different algorithms in Speaker and Mockingboard.

For both the buffer is 16*1024 samples = 371ms

1. Speaker (normal speed): `nNumSamplesError` is computed to target between 1/4 and 1/2 free space in the buffer (between 92ms and 185ms), but the number of samples generated is not affected. This value changes `g_nCpuCyclesFeedback`

https://github.com/AppleWin/AppleWin/blob/96a386a77a16f3d0a46f99472d1779e26fe89328/source/Speaker.cpp#L760

2. Speaker (full speed): the number of samples generated is changed to target < 92ms free space in the queue.

https://github.com/AppleWin/AppleWin/blob/96a386a77a16f3d0a46f99472d1779e26fe89328/source/Speaker.cpp#L559

3. Mockingboard: a hybrid: similar target as Speaker, but it directly affects the number of samples generated.

Am I understanding correctly that for Speaker it is the audio that drives the emulator speed, while for Mockingboard it is the cpu speed which drives the audio generation?

Speaker (case 1) If there are not enough bytes remaining, this means too much audio has been generated vs the cpu time executed, so we speed up the CPU which will execute a bit more than the standard 1ms worth of cycles next time?

For the other 2 cases, there is some "flexibility" on how much audio is possible to generate and the adjustment happens in the opposite direction.

Is this correct?

**Comment by tomcw** [comment]
Yes, that's largely correct.

>for Speaker it is the audio that drives the emulator speed

Yes, and it should quickly reach a steady state where the speaker audio doesn't need to influence (ie. add/subtract cycles from) the core 1MHz emulation speed.

>for Mockingboard it is the cpu speed which drives the audio generation

Yes... and for Mockingboard, if there's extra sample that need generating (if the buffer is too empty) then it's easy to generate more data to "top it up", since Mockingboard just generates a predictable waveform.

**Comment by audetto** [comment]
So, basically

44/45 is the "standard number of samples"

total error adjustment = 200, which means we can adjust from 44-200 = 0MHz, to 44+200 = 5.5MHz, with increments of 20 = 0.5 MHz.

Let's say the audio went at 3x the correct speed, we could still accelerate the CPU to handle it, but 10x it is not possible.

Since in Linux, I actually run 1 frame at a time, I better rescale these value to make sense.
I misunderstood it all, and I was targeting a precise CPU speed, as opposed to a precise audio speed, which is more important.

**Comment by audetto** [comment]
Question on Mockingboard:

why does the code get here even when no MB is installed?

https://github.com/AppleWin/AppleWin/blob/96a386a77a16f3d0a46f99472d1779e26fe89328/source/MockingboardCardManager.cpp#L229-L241

Should it first check if any MB is installed?

**Comment by tomcw** [comment]
It could do... but tracing the current code:
`UpdateSoundBuffer()` calls `GenerateAllSoundData()` which returns `numSamples = 0`.

So it's harmless as it stands.

But if you want to make `Update()` more intuitive, then we could change things so that for-loop also counts the number of Mockingboards, and then just early returns if there are none.

**Comment by audetto** [comment]
The only drawback is that it opens an entire audio device whose buffer is never filled.
I am revisiting the linux audio and I get 100% (harmless) underruns on this one.

Will send a PR to avoid it if possible.

**Comment by tomcw** [comment]
On Windows, I fill the DirectSound buffer with zeros (ie. silence), and then as it's a ring buffer, it just happily plays silence continuously - ie. no buffer underruns.

**Comment by audetto** [comment]
Yes, because the buffer is all zeros. But then, when the read pointer overtakes the write points, it is not clear if it is a real underrun or a harmless underrun.

This is in comparison with the buffer for the speaker or the same mockingboard when it really plays something.

**Comment by audetto** [comment]
Another question:

since the final speed is driven by the audio, then the exact speed at which we run the emulator is 23 * 44,100 = 1,014,300 Hz regardless of the theoretical speed for NTSC v PAL.

This is cause by the rounding happening in https://github.com/AppleWin/AppleWin/blob/47841bf58a8d5d298338b922bf1490b7cb02ef5f/source/Speaker.cpp#L185

which is 23 in all cases.

It means as well that the variable `g_fCurrentCLK6502` is actually misleading and that `g_nErrorInc` is constantly battling against it.

Would it not be easier to fix the CPU speed so that systematic audio correction would not be necessary (i.e. 1,014,300) ? And forget the NTSC / PAL value?
